src/hotspot/share/runtime/flags/jvmFlag.hpp

```diff
@@ -290,6 +291,27 @@ class JVMFlag {
 
   static const char* flag_error_str(JVMFlag::Error error);
 
+  // type checking
+  template <typename T>
+  static void assert_compatible_type(int flag_enum) {
+#ifndef PRODUCT
+    if (std::is_integral<T>::value) {
+      switch (flag_enum) {
+      case TYPE_bool:     assert(sizeof(T) == sizeof(bool)     && std::is_signed<T>::value == std::is_signed<bool>    ::value, "must be"); break;
+      case TYPE_int:      assert(sizeof(T) == sizeof(int)      && std::is_signed<T>::value == std::is_signed<int>     ::value, "must be"); break;
+      case TYPE_uint:     assert(sizeof(T) == sizeof(uint)     && std::is_signed<T>::value == std::is_signed<uint>    ::value, "must be"); break;
+      case TYPE_intx:     assert(sizeof(T) == sizeof(intx)     && std::is_signed<T>::value == std::is_signed<intx>    ::value, "must be"); break;
+      case TYPE_uintx:    assert(sizeof(T) == sizeof(uintx)    && std::is_signed<T>::value == std::is_signed<uintx>   ::value, "must be"); break;
+      case TYPE_uint64_t: assert(sizeof(T) == sizeof(uint64_t) && std::is_signed<T>::value == std::is_signed<uint64_t>::value, "must be"); break;
+      case TYPE_size_t:   assert(sizeof(T) == sizeof(size_t)   && std::is_signed<T>::value == std::is_signed<size_t>  ::value, "must be"); break;
+      default: ShouldNotReachHere();
+      }
+    } else {
+      assert(flag_enum == JVMFlag::TYPE_double, "must be double");
+    }
+#endif
+  }
+
```

Comment:
Do we need this? Seems an unused duplicate to the version in jvmFlagLimit.hpp.

Reply:
I left `JVMFlagLimit::assert_compatible_type` by mistake, so I deleted it. I think we can use `JVMFlag::assert_compatible_type` in other places as well to get rid of the 8x duplicated boilerplate functions such as `JVMFlag::{get_bool, get_int, get_intx, etc}`. I'll investigate and file an RFE.

For your other comments, I updated `JVMFlag::assert_compatible_type` to use macros, so there's less repetition and more uniform (no more special case for `double`). It's a little easier to maintain, but maybe not as obvious to read.